File: src/com/sterndu/json/JsonValue.java
package com.sterndu.json;

import java.util.function.Function;

public interface JsonValue {
	default String toJson() {
		return toJsonValue().toJson();
	}

	default String toJson(Function<Object, String> function) {
		return toJsonValue(function).toJson();
	}

	/**
	 *
	 * @return a JsonValue that is not this. Meant for Sub-Classes
	 */
	JsonValue toJsonValue();

	/**
	 *
	 * @return a JsonValue that is not this. Meant for Sub-Classes
	 */
	JsonValue toJsonValue(Function<Object, String> function);
}

File: src/com/sterndu/json/JsonValue.kt
@file:JvmName("JsonValue")
package com.sterndu.json

interface JsonValue {
	fun toJson(): String {
		return toJsonValue().toJson()
	}

	fun toJson(function: (Any?) -> String = { it.toString() }): String {
		return toJsonValue(function).toJson()
	}

	/**
	 *
	 * @return a JsonValue that is not this. Meant for Sub-Classes
	 */
	fun toJsonValue(): JsonValue

	/**
	 *
	 * @return a JsonValue that is not this. Meant for Sub-Classes
	 */
	fun toJsonValue(function: (Any?) -> String = { it.toString() }): JsonValue
}
